sycl/doc/extensions/DiscardQueueEvents/SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS.asciidoc

```diff
@@ -0,0 +1,223 @@
+
+  Queue.wait();
+----
+
+In the example above, the application doesn't use sycl events: `e1`, `e2`, `e3`
+and is waiting for the end of work by `queue::wait()`. The returned events will be
+_invalid_ events. _invalid_ event is a `sycl::event` with a status `invalid` and the event returns from each kernel
+submission when queue is created with the `discard_queue_events` property.
+See the description of behavior for this event below for details.
+
+_invalid_ event API behavior:
+[source,c++]
+----
+// the behavior is the same as for the default event, i.e. must throw an exception with the errc::invalid_object error code
+cl_event get() const
+
+// the behavior is the same as for the default event, i.e. must return true
+bool is_host() const;
+
+// the behavior is the same as for the default event, i.e. must return backend::host
+backend get_backend() const noexcept;
+
+// the behavior is the same as for the default event, i.e. must return a zero-sized vector
+std::vector<event> get_wait_list();
+
+// must throw an exception with the errc::invalid_object error code.
+void wait();
+
+// passing invalid event in the function must throw an exception with the errc::invalid_object error code.
+static void wait(const std::vector<event> &eventList);
+
+// must throw an exception with the errc::invalid_object error code.
+void wait_and_throw();
+
+// passing invalid event in the function must throw an exception with the errc::invalid_object error code.
+static void wait_and_throw(const std::vector<event> &eventList);
+
+// returns info::event_command_status::ext_oneapi_unknown status
+get_info<info::event::command_execution_status>() const;
+
+// the behavior is the same as for the default event, i.e. returns 0
+get_info<info::event::reference_count>() const;
+
+// must throw an exception with the errc::invalid_object error code.
+template <info::event_profiling param>
+typename param::return_type get_profiling_info() const;
+----
+
+The behavior when passing _invalid_ event to handler API:
+[source,c++]
+----
+// must throw an exception with the errc::invalid_object error code.
+handler::depends_on(event Event)
+
+// must throw an exception with the errc::invalid_object error code.
+handler::depends_on(const std::vector<event> &Events)
+----
+
+== Limitations
+
+The `discard_queue_events` property should be taken as a hint for SYCL implementation.
+In all kernel cases, SYCL implementation returns _invalid_ event, regardless of whether the optimization was done.
+Optimization occurs if the queue has `discard_queue_events` property and all features presented below are not met:
+
+ - Using out-of-order queue or using together with `enable_profiling` property
+ - Using https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc[fallback assert feature]
+ - Using streams, buffer/image accessors (excluding local accessors)
+

```

Comment:
Shouldn't we mention reductions here?

Reply:
I'm not sure if this needs to be written explicitly, because that case is already covered by using accessors. If you use accessors for reduction then it is already limited.

Reply:
I thought we support USM based reductions as well.
Tagging @v-klochkov

Reply:
> I thought we support USM based reductions as well.

yes, It is correct!


Accessor based reduction is NOT supported.
USM based reductions is supported.

Reply:
Doesn't the USM-based reduction use events behind the scenes, though?  A reduction might be enqueued as multiple kernels, and I believe that events/accessors are used as an implementation detail to keep those kernels ordered.  I don't think there's a special-case for reductions submitted to in-order queues.  @v-klochkov?

Reply:
If I understood it correctly, call of event::wait would cause exception, but reduction does not call wait().
Reduction implementation may create more than one kernel, and it connects them using either accessors (which has read after write sync builtin), or using depends_on(), which should work correctly with the new proposed attribute (e.g.: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/reduction.hpp#L1953)

IMO, reduction should keep working for both USM and buffer/accessors.